File: app/src/main/java/me/ykrank/s1next/widget/glide/AppHttpUrlLoader.java
package me.ykrank.s1next.widget.glide;

import androidx.annotation.Nullable;

import com.bumptech.glide.integration.okhttp3.OkHttpStreamFetcher;
import com.bumptech.glide.integration.okhttp3.OkHttpUrlLoader;
import com.bumptech.glide.load.Options;
import com.bumptech.glide.load.data.DataFetcher;
import com.bumptech.glide.load.model.GlideUrl;
import com.bumptech.glide.load.model.ModelLoader;
import com.bumptech.glide.load.model.ModelLoaderFactory;
import com.bumptech.glide.load.model.MultiModelLoaderFactory;
import com.github.ykrank.androidtools.widget.glide.model.ForcePassUrl;

import org.jetbrains.annotations.NotNull;

import java.io.InputStream;

import javax.inject.Inject;

import me.ykrank.s1next.App;
import me.ykrank.s1next.data.pref.DownloadPreferencesManager;
import me.ykrank.s1next.widget.glide.model.AvatarUrl;
import okhttp3.OkHttpClient;

/**
 * A simple model loader for fetching media over HTTP/HTTPS using OkHttp.
 * <p>
 * Forked from {@link OkHttpUrlLoader}
 */
final public class AppHttpUrlLoader implements ModelLoader<GlideUrl, InputStream> {

    private final OkHttpClient mOkHttpClient;

    private AppHttpUrlLoader(OkHttpClient okHttpClient) {
        this.mOkHttpClient = okHttpClient;
    }

    @Override
    public LoadData<InputStream> buildLoadData(@NotNull GlideUrl glideUrl, int width, int height, @NotNull Options options) {
        return new LoadData<>(glideUrl, buildResourceFetcher(glideUrl, width, height, options));
    }

    @Override
    public boolean handles(@NotNull GlideUrl glideUrl) {
        return true;
    }

    private DataFetcher<InputStream> buildResourceFetcher(GlideUrl model, int width, int height, Options options) {
        if (model instanceof AvatarUrl) {
            return new AvatarStreamFetcher(mOkHttpClient, (AvatarUrl) model);
        } else if (model instanceof ForcePassUrl) {
            return new MultiThreadHttpStreamFetcher(mOkHttpClient, model);
        }
        return new AppHttpStreamFetcher(mOkHttpClient, model);
    }

    /**
     * The default factory for {@link AppHttpUrlLoader}.
     */
    public static final class Factory implements ModelLoaderFactory<GlideUrl, InputStream> {

        private final OkHttpClient mOkHttpClient;

        /**
         * Constructor for a new Factory that runs requests using given client.
         */
        public Factory(OkHttpClient okHttpClient) {
            this.mOkHttpClient = okHttpClient;
        }

        @Override
        public ModelLoader<GlideUrl, InputStream> build(@NotNull MultiModelLoaderFactory multiFactory) {
            return new AppHttpUrlLoader(mOkHttpClient);
        }

        @Override
        public void teardown() {
            // Do nothing, this instance doesn't own the client.
        }
    }
}

File: app/src/main/java/me/ykrank/s1next/widget/glide/AppHttpUrlLoader.kt
package me.ykrank.s1next.widget.glide

import com.bumptech.glide.integration.okhttp3.OkHttpStreamFetcher
import com.bumptech.glide.load.Options
import com.bumptech.glide.load.data.DataFetcher
import com.bumptech.glide.load.model.GlideUrl
import com.bumptech.glide.load.model.ModelLoader
import com.bumptech.glide.load.model.ModelLoaderFactory
import com.bumptech.glide.load.model.MultiModelLoaderFactory
import com.github.ykrank.androidtools.widget.glide.model.ForcePassUrl
import me.ykrank.s1next.widget.glide.model.AvatarUrl
import okhttp3.OkHttpClient
import java.io.InputStream

/**
 * A simple model loader for fetching media over HTTP/HTTPS using OkHttp.
 *
 *
 * Forked from [OkHttpUrlLoader]
 */
class AppHttpUrlLoader private constructor(
    private val mOkHttpClient: OkHttpClient,
    private val mProgressOkHttpClient: OkHttpClient
) :
    ModelLoader<GlideUrl, InputStream> {
    override fun buildLoadData(
        glideUrl: GlideUrl,
        width: Int,
        height: Int,
        options: Options
    ): ModelLoader.LoadData<InputStream>? {
        return ModelLoader.LoadData(
            glideUrl,
            buildResourceFetcher(glideUrl, width, height, options)
        )
    }

    override fun handles(glideUrl: GlideUrl): Boolean {
        return true
    }

    private fun buildResourceFetcher(
        model: GlideUrl,
        width: Int,
        height: Int,
        options: Options
    ): DataFetcher<InputStream> {
        if (model is AvatarUrl) {
            return AvatarStreamFetcher(mOkHttpClient, model)
        } else if (model is ForcePassUrl) {
            return OkHttpStreamFetcher(mProgressOkHttpClient, model)
        }
        return AppHttpStreamFetcher(mOkHttpClient, model)
    }

    /**
     * The default factory for [AppHttpUrlLoader].
     */
    class Factory(
        private val mOkHttpClient: OkHttpClient,
        private val mProgressOkHttpClient: OkHttpClient
    ) :
        ModelLoaderFactory<GlideUrl, InputStream> {
        override fun build(multiFactory: MultiModelLoaderFactory): ModelLoader<GlideUrl, InputStream> {
            return AppHttpUrlLoader(mOkHttpClient, mProgressOkHttpClient)
        }

        override fun teardown() {
            // Do nothing, this instance doesn't own the client.
        }
    }
}
